### DISCUSSION TITLE
"405 Method Not Allowed" Error when trying to make a POST request to my "/url" endpoint hosted on Vercel

### DISCUSSION THREAD

**Discussion by L1xus** [vercel/community#3911] [Help] [answered]
### Summary

My frontend works perfectly fine but I keep getting errors when I'm interacting with my vercel server even though that everything works fine in my localhost !!!

Any help please I've been struggling with this for DAAAYS ;/

### Steps to Reproduce

My project Structure is like this :
```
>node_modules
>public
   >css
   >js
   index.html
package.json
index.js

```
I've deployed everything on my vercel project and I've specified  the routing correctly on vercel.json file
```
{
    "rewrites": [
        {
            "source": "/url", 
            "destination": "index.js"
        }
    ]
}
``` 
here is the index.js code that receives post requests
```
const app = express();
const port =  443 || 80 || 3000;

app.use(express.static(__dirname + '/public'));
app.use(bodyParser.json());

app.post('/url', async (req, res) => {
    console.log('Received POST request at /url');
    let { videoUrl } = req.body;
    
    // get url video
....
....
....
``` 

and here is the js code that sends the requests:
 
```
export const getUrl = async (url) => {
    let content = document.getElementById('content');
    console.log('Client-side sending request to /url');
    let res = await fetch('/url', {
        method: 'POST',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ videoUrl: url })
    })

    let { vid } = await res.json();

....
....
``` 
My frontend works perfectly fine but I keep getting errors when I'm interacting with my vercel server even though that everything works fine in my localhost !!!

Any help please I've been struggling with this for DAAAYS ;/

**Comment by L1xus** [comment]
@amyegan @swarnava can you please check this out !

**Comment by TooTallNate** [comment, marked as answer]
You should put the `index.js` file in the `api` directory instead. That will make it be executed as a Node.js function. Putting it at the root-level like you are doing now will make it be a static file.

In your client-side code, you can invoke the endpoint at `POST /api` (index is implied as the default). So there is no need for the rewrite in `vercel.json`.

**Reply by L1xus** [reply]
thank you so much for your response!
So my project structure should be like this ?

```
-> api 
     -index.js
->node_modules 
->public
    ->css
    ->js
    - index.html
-package.json

```
and I have to change my endpoint to /api in my client side code and  without the need for vecel.json ?

**Reply by KickButtowski80** [reply]
what would happen if the name of the file was something else rather than index.js?
this issue just happen to me in development FYI

**Reply by sigmarsson** [reply]
Having perhaps a clue also in the case of RSC as depicted beneath ? I encounter 405 as 

<img width="863" alt="image" src="https://github.com/vercel/community/assets/43340959/915dfa00-c0ff-43c3-b79d-668049d99b6b">


<img width="215" alt="image" src="https://github.com/vercel/community/assets/43340959/c99141ab-a668-4bd0-8455-2023fd7c55f8">
<img width="522" alt="image" src="https://github.com/vercel/community/assets/43340959/5045a655-952c-4dc9-8042-db24cddbf1da">

**Reply by sigmarsson** [reply]
and btw, why console.logs do not appear on Vercel's runtime log ?

**Reply by watsonwebdesign** [reply]
Client-side console.logs will only appear on the browser. Server-side will appear in Vercel's runtime log. So if you're logging in a client component, it will not be stored.

[a comment by an automated account is not shown]
